### PR TITLE
[FIXED JENKINS-28506] - Git fails to trigger from polling for refs/heads/ style branch name

### PR DESCRIPTION
**Pull request by ndeloof**
Proposed alternative to https://github.com/jenkinsci/git-plugin/pull/323

the actual fix is to split BranchSpec matching check in two steps:
- first assume user did used a fully qualified refspec `refs/(heads|tags)/*`
- then convert the remote head reference into short notation `remote/branch`

### CODE REVIEW

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 1]
```diff
@@ -583,22 +583,28 @@ private PollingResult compareRemoteRevisionWithImpl(Job<?, ?> project, Launcher
...
-                            else name = remote + "/" + head;
 
-                            if (!branchSpec.matches(name, environment)) continue;
+                            // first, check the a canonical git reference is configured
+                            if (!branchSpec.matches(head, environment)) {
+
+                                // convert head `refs/*` into shortcut notation `remote/branch`
+                                String name = head;
+                                if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
+                                else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Again hardcode, this wouldn't work with `ref/notes` or any other git based solution. `refs/heads` is just a DEFAULT for git branches.

I can propose add new UI specifier as "branch specifier", but some 'git refspec specifier' that will work right as git documentation, move matching to this class.

Comment (ndeloof) [inline src/main/java/hudson/plugins/git/GitSCM.java]: So you suggest to match `refs/*/` ? I never heard about refs/notes usage with a jenkins job, do you know a use-case for this ?

Also, Branch Specifier help tips only claim to support refs/heads/<branchName>, refs/remotes/<remoteRepoName>/<branchName> and refs/tags/<tagName>, so anything else to actually work would be only by chance

I'd like such a stricter approach you suggest, need to be carreful with legacy config conversion

Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/GitSCM.java]: @ndeloof you can even do github PR builds without any PR plugin, github stores all PRs in refs/pull/$ID/{head,merge} . This is not documented but this is how all gh plugins works https://github.com/jenkinsci/github-pullrequest-plugin/blob/master/README.md (Configuration part).

Also https://github.com/jenkinsci/github-plugin/pull/39 probably unrelated, but in general refs hardcodes raising many questions.

Comment (ndeloof) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Yes I know ghprb plugin relies on refs/pull, but this require some additional refspec configured (admittedly not used here, should be considered). I never said this is feature complete, just want JENKINS-28506 fixed, future improvements are still possible to offer more flexibility

**Comment by ndeloof** [comment]
@MarkEWaite all test pass now :D but as you know git-plugin test suite do only check the tip of the iceberg, so a finer review is welcome. also @reviewbybees

File: pom.xml [hunk 2]
```diff
@@ -263,7 +263,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>matrix-project</artifactId>
-      <version>1.4.1</version>
+      <version>1.4</version>
```
Comment (oleg-nenashev) [inline pom.xml]: Hm? Is it related somehow?

Comment (oleg-nenashev) [inline pom.xml]: Never mind, I've finally read the Wiki. This Matrix Project thing should have been released as 2.0 at least

Comment (ndeloof) [inline pom.xml]: you just can't run mvn hpi:run with 1.4.1 - make it hard to test intermediate commits :P

**Comment by MarkEWaite** [comment]
I won't return from vacation for another week. Could you check with the submitter of the original bug if he's willing to test a pre-release of the plugin before I return?

Does this need a matching pre-release of the git client plugin (some form of 1.18.0 pre-release)? The ISO date format parsing bug is also resolved in this, but will it interact badly with the changelog format written by git client plugin 1.16.1?

File: pom.xml (hunk 2 above)
Comment (MarkEWaite) [inline pom.xml]: Would it be better to drop this change from this commit and instead merge https://github.com/jenkinsci/git-plugin/pull/322 ?  That would remove one more pull request from the backlog and make this change explicit

Is there a matching change which will allow the JUnit plugin to work with 1.580 (see that preceding pull request for Kohsuke's reference to the JUnit special needed)

**Comment by ndeloof** [comment]
removed 451c47b, rebased for cleanup of fixup commits

**Pull request merged**